### PR TITLE
feat: add type hints to core lib

### PR DESCRIPTION
**Pull request by tushar5526**
Related to #40

### CODE REVIEW

**Comment by tushar5526** [comment]
Hey @khvn26 @matthewelwell can you run the workflows, please :)

**Comment by dabeeeenster** [comment]
Thanks so much! one of the team will take a look

File: flagsmith/flagsmith.py
Comment (zachaysan) [inline flagsmith/flagsmith.py:11]: Just out of curiosity why was `type: ignore` necessary here?

File: flagsmith/flagsmith.py
Comment (zachaysan) [inline flagsmith/flagsmith.py:26]:
```suggestion
DEFAULT_API_URL: typing.Final[str] = "https://edge.api.flagsmith.com/api/v1/"
```

File: flagsmith/flagsmith.py
Comment (zachaysan) [inline flagsmith/flagsmith.py:52]: Why did you reorder this (and other) argument(s) here? I'm sure people call this with kwargs only, but I don't really see an advantage of reordering them without necessity.

File: flagsmith/flagsmith.py
Comment (zachaysan) [inline flagsmith/flagsmith.py:216]: Nit: `self._environment` is probably `None` but it would be better to return the type explicitly like this:
```python
raise TypeError(f"Environment should not be '{type(self._environment)}'")
```

File: flagsmith/flagsmith.py
Comment (zachaysan) [inline flagsmith/flagsmith.py:228]: Same idea. May be worth while to make a single private method to raise the type error if present. Not sure.

File: tests/test_flagsmith.py
Comment (zachaysan) [inline tests/test_flagsmith.py:104]: I would suggest a short comment here why an `isinstance` check is necessary here.

**Review by zachaysan — COMMENTED** [review]
Hey @tushar5526 thanks a lot for putting this review together.

I'm on the Flagsmith team, but I haven't touched this repository before, so I can't run the workflows you requested from Kim and Matt. They're heads down for a project for the next week, but I'll see if someone else can run the workflows and maybe we'll get lucky.

I've added a bunch of review comments, but please keep in mind that I don't have full visibility on this project. 

Feel free to push back on any of them that don't make sense 😄

File: flagsmith/flagsmith.py
Comment (tushar5526) [reply flagsmith/flagsmith.py:52]: Good catch, I was experimenting with whether this really needs a default value or not, and forgot to re-arrange the args back. I will fix this. Thanks!

File: flagsmith/flagsmith.py
Comment (tushar5526) [reply flagsmith/flagsmith.py:228]: Agreed, I will see if a `TypeGuard` is what I need here.

File: tests/test_flagsmith.py
Comment (tushar5526) [reply tests/test_flagsmith.py:104]: Sure!

File: flagsmith/flagsmith.py
Comment (tushar5526) [reply flagsmith/flagsmith.py:216]: Yes, agreed. Thanks!

File: flagsmith/flagsmith.py
Comment (tushar5526) [reply flagsmith/flagsmith.py]: Update the `type: ignore` to be more specific.

Comment (khvn26) [reply flagsmith/flagsmith.py]: @tushar5526 Could you change the import to `flag_engine.identities.traits.TraitModel` instead?

Comment (tushar5526) [reply flagsmith/flagsmith.py]: Done, mypy is not throwing any errors. Marking this as resolved.

File: flagsmith/flagsmith.py
Comment (khvn26) [reply flagsmith/flagsmith.py]: I'll take a closer look later this week but I'm not sure this is at all needed? A simple check for `None` should suffice here. This code is probably a leftover (or just bad Python code).

File: flagsmith/__init__.py
Comment (khvn26) [inline flagsmith/__init__.py:3]: 1. I believe that, with the explicit export added, the `# noqa` comment can be removed now.

2. Nitpick: I prefer using tuples for the exports as there's no sense in using mutable sequences for those.

File: flagsmith/flagsmith.py
Comment (khvn26) [inline flagsmith/flagsmith.py:28]: Is this function still used?

File: flagsmith/flagsmith.py
Comment (khvn26) [inline flagsmith/flagsmith.py:44]: If this stays, I'd prefer to have it in the `utils` module.

File: flagsmith/flagsmith.py
Comment (khvn26) [inline flagsmith/flagsmith.py:228]: Again, this `type: ignore` should go away if we opt for a simple `None` check.

File: flagsmith/polling_manager.py
Comment (khvn26) [inline flagsmith/polling_manager.py:19]: Please clarify the error you're ignoring here.

**Review by khvn26 — CHANGES_REQUESTED** [review]

File: flagsmith/flagsmith.py
Comment (tushar5526) [reply flagsmith/flagsmith.py]: Sure, I believe `None` is better here. `TypeGuards` now seems much more suitable when there are different types involved 🤔

File: flagsmith/flagsmith.py
Comment (tushar5526) [reply flagsmith/flagsmith.py:28]: No, will remove this. Thanks!

File: flagsmith/__init__.py
Comment (tushar5526) [reply flagsmith/__init__.py:3]: Removed #noqa, also I tried switching to tuples, but mypy is erroring out and not recognising the exports for some reason.

File: flagsmith/flagsmith.py
Comment (tushar5526) [reply flagsmith/flagsmith.py:228]:
```py
def _get_environment_flags_from_document(self) -> Flags:
    if self._environment:
        return Flags.from_feature_state_models(
            feature_states=engine.get_environment_feature_states(self._environment),
            analytics_processor=self._analytics_processor,
            default_flag_handler=self.default_flag_handler,
        )
```

Using `None` still gives a `mypy` error. I have to probably raise an error in the else clause to fix the `no return` error but that looks like additional noise.

File: flagsmith/polling_manager.py
Comment (tushar5526) [reply flagsmith/polling_manager.py:19]: I was ignoring `*args` and `**kwargs`, added typing for it.

File: flagsmith/__init__.py
Comment (khvn26) [reply flagsmith/__init__.py]: Have you by chance omitted a trailing comma?

```python
# assigns a tuple
__all__ = ("Flagsmith",)

# assigns a string :)
__all__ = ("Flagsmith")
```

File: flagsmith/flagsmith.py
Comment (khvn26) [reply flagsmith/flagsmith.py]: This is fine by me:

```python
if self._environment is None:
    raise TypeError('No environment present')
```

File: flagsmith/__init__.py
Comment (tushar5526) [reply flagsmith/__init__.py]: Ahh, my bad 😣 Yes I missed using a trailing comma! Fixed, thanks!

File: flagsmith/flagsmith.py
Comment (tushar5526) [reply flagsmith/flagsmith.py]: Added

File: flagsmith/utils/identities.py
Comment (khvn26) [inline flagsmith/utils/identities.py:6]: Let's use a `TypedDict` for this.

File: flagsmith/flagsmith.py
Comment (khvn26) [inline flagsmith/flagsmith.py:45]: I _think_ header value type should just be `str`.

File: flagsmith/flagsmith.py
Comment (khvn26) [inline flagsmith/flagsmith.py:157]: Wouldn't hurt to use the `flag_engine.identities.traits.types.TraitValue` type for the trait values here.

File: flagsmith/flagsmith.py
Comment (khvn26) [inline flagsmith/flagsmith.py:178]: Wouldn't hurt to use the flag_engine.identities.traits.types.TraitValue type for the trait values here.

File: flagsmith/flagsmith.py
Comment (khvn26) [inline flagsmith/flagsmith.py:218]: Wouldn't hurt to use the flag_engine.identities.traits.types.TraitValue type for the trait values here.

File: flagsmith/flagsmith.py
Comment (khvn26) [inline flagsmith/flagsmith.py:275]: 1. Our SDK endpoints only return JSON objects, so it'll be fair to use a `typing.Dict[str, Any]` for the return type annotation. This will probably require a `cast` but, IMO, will improve developer experience.
2. Maybe it makes sense to define a recursive `JsonType = Union[None, int, str, bool, List[JsonType], Dict[str, JsonType]]` type and use it both for the `body` argument and the return annotation (`Dict[str, JsonType]`).

File: flagsmith/flagsmith.py
Comment (khvn26) [inline flagsmith/flagsmith.py:293]: Wouldn't hurt to use the flag_engine.identities.traits.types.TraitValue type for the trait values here.

File: flagsmith/models.py
Comment (khvn26) [inline flagsmith/models.py:15]: `type[None]` can be just `None`? Actually I'm not even sure all Python versions we support implement `__getitem__` for this buit-in.

**Review by khvn26 — CHANGES_REQUESTED** [review]

**Comment by khvn26** [comment]
@tushar5526 Let us know if you need help merging this — we're working on a task involving the Python client and would love to include typing in the next release.

**Comment by tushar5526** [comment]
@khvn26 thanks, I will probably get sometime this weekend before I work on this :)

File: flagsmith/models.py
Comment (tushar5526) [reply flagsmith/models.py:15]: Yes, `type[None]` is supported by python 3.9+. Using just `None`

File: flagsmith/flagsmith.py
Comment (tushar5526) [reply flagsmith/flagsmith.py:275]: ```JsonType = Union[None, int, str, bool, List[JsonType], Dict[str, JsonType]]``` - We would have to add asserts in the current code to narrow down types as `response[key]` is failing because `response` could be a string as well. We can define a definite type for the response API as that might be helpful for developers. What do you think?

File: flagsmith/models.py
Comment (tushar5526) [reply flagsmith/models.py]: We will have to specify `is_default` as a `kw_only` field to set default values in parent dataclass child classes. `kw_only` is only available in python 3.10+ . It seems using `_init_ ` is the only possible way to do this for python < 3.10

https://stackoverflow.com/questions/49908182/how-to-make-keyword-only-fields-with-dataclasses

**Comment by tushar5526** [comment]
Hey @khvn26 @matthewelwell Happy New Year! Need some help with Python 3.8 and 3.9. Got an issue with inheritance in dataclasses. `Flag` and `DefaultFlag` are child dataclasses inheriting from `BaseFlag` dataclass. To override the `is_default` member with a default value we would have to use `kw_only`, but it's only available in Python >= 3.10. Couldn't find a workaround for lower versions. Thinking we might have to stick with using __init__ in subclasses (as done earlier in code). Any ideas or workarounds that you folks are aware of?

**Comment by khvn26** [comment]
Hey @tushar5526! The good solution to this is to define the field individually in the subclasses. I've done it in my PR here: https://github.com/Flagsmith/flagsmith-python-client/pull/72/files#diff-90b21583a0688d12b6b9a8d2970969907e7d49546b5f953ac1cfb61d14111518

The PR I've linked to is ready to merge. I suggest you rebase #71 on the main branch after #72 is merged. Other than that, there's [one outstanding comment](https://github.com/Flagsmith/flagsmith-python-client/pull/71/files#r1481786019) that needs resolving. After that we'll finally merge your PR 👍

File: flagsmith/models.py
Comment (tushar5526) [reply flagsmith/models.py:50]: I tried adding a JsonType here but got a bunch of errors. If we want to do something like `flag_data["key"]["nested-key"]` then we will have to narrow the type up to 2 levels of nesting. Narrow the type to dict for `flag_data["key"]` and then narrow it further to dict for `flag_data["key"]["nested-key"]` as JsonType could be any of `int | None | str | JsonType ...`

Hovering over JsonType reveals a sizable recursive type hint. Would it be clearer for users if we defined types reflecting the structure of responses from various endpoints instead of relying on a generic JsonType?

**Comment by matthewelwell** [comment]
@tushar5526 apologies for the delay on this but could you take a look at the failing unit tests or would you like one of us to pick it up?

**Comment by tushar5526** [comment]
@matthewelwell I forgot about this, was waiting for #72 to be merged. I believe I can rebase on top of it and fix the tests.

**Comment by matthewelwell** [comment]
@khvn26 can we merge #72 ?

**Comment by khvn26** [comment]
@matthewelwell @tushar5526 #72 merged! Feel free to rebase and fix stuff!

File: flagsmith/models.py
Comment (khvn26) [reply flagsmith/models.py:43]: This is a great point Tushar. Unfortunately, our API is not quite ready yet. We'll work on out v2 API that will export contracts suitable for code generation soo!. For now, we have to stick with generic types.

**Comment by tushar5526** [comment]
Fixed!

File: flagsmith/flagsmith.py
Comment (tushar5526) [reply flagsmith/flagsmith.py]: I went ahead with using `any` as _get_json_reponse() also returns a `list` of `dicts`. Using `any` and type-casting the response later worked fine.

**Comment by tushar5526** [comment]
Fixed the conflicts.

**Pull request merged**
